data_model/derive/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+//! Crate with `Filter` and `EventFilter` derive macro
+#![allow(clippy::expect_used, clippy::panic, clippy::too_many_lines)]
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+
+/// [`Filter`] is used for code generation of `...Filter` structs and `...EventFilter` enums, as well as
+/// implementing the `Filter` trait for both of them. This macro should only be attributed to `Event` enums.
+/// Due to name scoping, the macro currently only
+/// expands properly from within the `iroha_data_model` crate as it relies on a few of `crate::prelude`
+/// imports. This macro is also dependent on the naming conventions adopted so far, such as that
+/// `Event` enums always have tuple variants with either some sort of `Id` or another `Event` inside
+/// of them.
+#[proc_macro_derive(Filter)]
+pub fn filter_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("Failed to parse input TokenStream.");
+    impl_from_filter(&ast)
+}
+
+fn impl_from_filter(ast: &syn::DeriveInput) -> TokenStream {
```

Comment:
Overall, I think this function should be broken up into smaller pieces. More `structs` could help you with parsing, so that illegal states (Events without the `Event` suffix) are unrepresentable. It's good enough for a start.